File: ch14/python/eventhub/eventhub_example.py
# Python example: Azure Event Hub - Send and Receive Event
# Requires: pip install azure-eventhub azure-identity

from azure.eventhub import EventHubProducerClient, EventData, EventHubConsumerClient
from azure.identity import DefaultAzureCredential

connection_string = "<your_connection_string>"
eventhub_name = "myeventhub"
fully_qualified_namespace = "<your_eventhub_namespace>.servicebus.windows.net"
consumer_group = "$Default"
message = "Hello from Python!"

# 1. Using connection string
producer_conn_str = EventHubProducerClient.from_connection_string(conn_str=connection_string, eventhub_name=eventhub_name)
with producer_conn_str:
    event_data_batch = producer_conn_str.create_batch()
    event_data_batch.add(EventData(message))
    producer_conn_str.send_batch(event_data_batch)

consumer_conn_str = EventHubConsumerClient.from_connection_string(conn_str=connection_string, consumer_group=consumer_group, eventhub_name=eventhub_name)
received = False
def on_event_conn_str(partition_context, event):
    global received
    print(f"Received (conn str): {event.body_as_str()}")
    received = True
    consumer_conn_str.close()

with consumer_conn_str:
    consumer_conn_str.receive(on_event=on_event_conn_str, starting_position="-1", max_wait_time=5)

# 2. Using Entra ID (DefaultAzureCredential)
producer_entra = EventHubProducerClient(fully_qualified_namespace=fully_qualified_namespace, eventhub_name=eventhub_name, credential=DefaultAzureCredential())
with producer_entra:
    event_data_batch = producer_entra.create_batch()
    event_data_batch.add(EventData(message))
    producer_entra.send_batch(event_data_batch)

consumer_entra = EventHubConsumerClient(fully_qualified_namespace=fully_qualified_namespace, eventhub_name=eventhub_name, consumer_group=consumer_group, credential=DefaultAzureCredential())
received = False
def on_event_entra(partition_context, event):
    global received
    print(f"Received (Entra ID): {event.body_as_str()}")
    received = True
    consumer_entra.close()

with consumer_entra:
    consumer_entra.receive(on_event=on_event_entra, starting_position="-1", max_wait_time=5)

File: ch14/java/eventhub/EventHubExample.java
// Java example: Azure Event Hub - Send and Receive Event
// Requires Maven: com.azure:azure-messaging-eventhubs, com.azure:azure-identity

import com.azure.messaging.eventhubs.*;
import com.azure.identity.DefaultAzureCredentialBuilder;
import java.util.Collections;

public class EventHubExample {
    public static void main(String[] args) {
        String eventHubName = "myeventhub";
        String connectionString = "<your_connection_string>";
        String fullyQualifiedNamespace = "<your_eventhub_namespace>.servicebus.windows.net";
        String message = "Hello from Java!";
        String consumerGroup = EventHubClientBuilder.DEFAULT_CONSUMER_GROUP_NAME;

        // 1. Using connection string
        EventHubProducerClient producerConnStr = new EventHubClientBuilder()
                .connectionString(connectionString, eventHubName)
                .buildProducerClient();
        producerConnStr.send(new EventData(message));
        producerConnStr.close();

        EventHubConsumerClient consumerConnStr = new EventHubClientBuilder()
                .connectionString(connectionString, eventHubName)
                .consumerGroup(consumerGroup)
                .buildConsumerClient();
        for (PartitionEvent event : consumerConnStr.receive(false, 1)) {
            System.out.println("Received (conn str): " + event.getData().getBodyAsString());
            break;
        }
        consumerConnStr.close();

        // 2. Using Entra ID (DefaultAzureCredential)
        EventHubProducerClient producerEntra = new EventHubClientBuilder()
                .fullyQualifiedNamespace(fullyQualifiedNamespace)
                .eventHubName(eventHubName)
                .credential(new DefaultAzureCredentialBuilder().build())
                .buildProducerClient();
        producerEntra.send(new EventData(message));
        producerEntra.close();

        EventHubConsumerClient consumerEntra = new EventHubClientBuilder()
                .fullyQualifiedNamespace(fullyQualifiedNamespace)
                .eventHubName(eventHubName)
                .consumerGroup(consumerGroup)
                .credential(new DefaultAzureCredentialBuilder().build())
                .buildConsumerClient();
        for (PartitionEvent event : consumerEntra.receive(false, 1)) {
            System.out.println("Received (Entra ID): " + event.getData().getBodyAsString());
            break;
        }
        consumerEntra.close();
    }
}
